core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -325,12 +325,11 @@ private synchronized Map<PlanNodeId, TaskSource> updateSources(List<TaskSource>
         Map<PlanNodeId, TaskSource> updatedUnpartitionedSources = new HashMap<>();
 
         // first remove any split that was already acknowledged
-        long currentMaxAcknowledgedSplit = this.maxAcknowledgedSplit;
         sources = sources.stream()
                 .map(source -> new TaskSource(
                         source.getPlanNodeId(),
                         source.getSplits().stream()
-                                .filter(scheduledSplit -> scheduledSplit.getSequenceId() > currentMaxAcknowledgedSplit)
+                                .filter(scheduledSplit -> scheduledSplit.getSequenceId() > maxAcknowledgedSplitByPlanNode.getOrDefault(source.getPlanNodeId(), Long.MIN_VALUE))
```

Comment:
There could be lots of splits. In this case, there will be lots of map accesses. It would be better to do like the below,

                .map(source -> {
                    long currentMaxAcknowledgedSplit = maxAcknowledgedSplitByPlanNode.getOrDefault(source.getPlanNodeId(), Long.MIN_VALUE);
                    return new TaskSource(
                            source.getPlanNodeId(),
                            source.getSplits().stream()
                                    .filter(scheduledSplit -> scheduledSplit.getSequenceId() > currentMaxAcknowledgedSplit)
                                    .collect(Collectors.toSet()),
                            // Like splits, noMoreSplitsForLifespan could be pruned so that only new items will be processed.
                            // This is not happening here because correctness won't be compromised due to duplicate events for noMoreSplitsForLifespan.
                            source.getNoMoreSplitsForLifespan(),
                            source.isNoMoreSplits());
                })